--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1919,29 +1919,25 @@ int do_execve_file(struct file *file, void *__argv, void *__envp)
 {
 	struct user_arg_ptr argv = { .ptr.native = __argv };
    struct user_arg_ptr envp = { .ptr.native = __envp };

    return __do_execve_file(AT_FDCWD, NULL, argv, envp, 0, file);
 }

#ifdef CONFIG_KSU
-extern bool ksu_execveat_hook __read_mostly;
-extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
-							   void *envp, int *flags);
-extern int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
-										 void *argv, void *envp, int *flags);
+__attribute__((hot))
+extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
+							   void *envp, int *flags);
#endif
+
 int do_execve(struct filename *filename,
 	const char __user *const __user *__argv,
 	const char __user *const __user *__envp)
 {
 	struct user_arg_ptr argv = { .ptr.native = __argv };
 	struct user_arg_ptr envp = { .ptr.native = __envp };
#ifdef CONFIG_KSU
-	if (unlikely(ksu_execveat_hook))
-        ksu_handle_execveat((int *)AT_FDCWD, &filename, &argv, &envp, 0);
+	ksu_handle_execveat((int *)AT_FDCWD, &filename, &argv, &envp, 0);
-   else
-		ksu_handle_execveat_sucompat((int *)AT_FDCWD, &filename, NULL, NULL, NULL);
#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
@@ -1961,6 +1961,5 @@ static int compat_do_execve(struct filename *filename,
 #ifdef CONFIG_KSU
-    if (!ksu_execveat_hook)
-       ksu_handle_execveat_sucompat((int *)AT_FDCWD, &filename, NULL, NULL, NULL); /* 32-bit support */
+	 ksu_handle_execveat((int *)AT_FDCWD, &filename, &argv, &envp, 0);
#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }


--- a/fs/open.c
+++ b/fs/open.c
@@ -454,8 +454,9 @@ long do_faccessat(int dfd, const char __user *filename, int mode)
 	return res;
 }

#ifdef CONFIG_KSU
+__attribute__((hot))
extern int ksu_handle_faccessat(int *dfd, const char __user **filename_user,
				int *mode, int *flags);
#endif

--- a/fs/stat.c
+++ b/fs/stat.c
@@ -353,6 +353,10 @@ SYSCALL_DEFINE2(newlstat, const char __user *, filename,
 	return cp_new_stat(&stat, statbuf);
 }

#ifdef CONFIG_KSU
+__attribute__((hot))
extern int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags);
#endif


--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ -310,6 +310,11 @@ static DEFINE_MUTEX(reboot_mutex);
  *
  * reboot doesn't sync: do that yourself before calling this.
  */
+
+#ifdef CONFIG_KSU
+extern int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg);
+#endif
+
 SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 		void __user *, arg)
 {
@@ -317,6 +317,9 @@ SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 	char buffer[256];
 	int ret = 0;

+#ifdef CONFIG_KSU
+	ksu_handle_sys_reboot(magic1, magic2, cmd, &arg);
+#endif
 	/* We only trust the superuser with rebooting the system. */
 	if (!ns_capable(pid_ns->user_ns, CAP_SYS_BOOT))
 		return -EPERM;

--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -594,8 +594,18 @@ ssize_t ksys_read(unsigned int fd, char __user *buf, size_t count)
 	return ret;
 }

#ifdef CONFIG_KSU
extern bool ksu_vfs_read_hook __read_mostly;
-extern int ksu_handle_sys_read(unsigned int fd, char __user **buf_ptr,
-			    size_t *count_ptr);
+extern __attribute__((cold)) int ksu_handle_sys_read(unsigned int fd,
+				char __user **buf_ptr, size_t *count_ptr);
#endif
+
SYSCALL_DEFINE3(read, unsigned int, fd, char __user *, buf, size_t, count)


--- a/drivers/input/input.c
+++ b/drivers/input/input.c
@@ -444,11 +444,22 @@ static void input_handle_event(struct input_dev *dev,
  * to 'seed' initial state of a switch or initial position of absolute
  * axis, etc.
  */
#ifdef CONFIG_KSU
extern bool ksu_input_hook __read_mostly;
-extern int ksu_handle_input_handle_event(unsigned int *type, unsigned int *code, int *value);
+extern __attribute__((cold)) int ksu_handle_input_handle_event(
+			unsigned int *type, unsigned int *code, int *value);
#endif
